x-pack/filebeat/input/azureeventhub/v2_input.go

```diff
@@ -0,0 +1,532 @@
+
+	// Initialize the components needed to process events,
+	// in particular the consumerClient.
+	err = in.setup(ctx)
+	if err != nil {
+		return err
+	}
+	defer in.consumerClient.Close(context.Background())
+
+	// Store a reference to the pipeline, so we
+	// can create a new pipeline client for each
+	// partition.
+	in.pipeline = pipeline
+
+	// Start the main run loop
+	in.run(ctx)
+
+	return nil
+}
+
+// setup initializes the components needed to process events.
+func (in *eventHubInputV2) setup(ctx context.Context) error {
+
+	// Decode the messages from event hub into
+	// a `[]string`.
+	in.messageDecoder = messageDecoder{
+		config:  in.config,
+		log:     in.log,
+		metrics: in.metrics,
+	}
+
+	// FIXME: check more pipelineClient creation options.
+	containerClient, err := container.NewClientFromConnectionString(
+		in.config.SAConnectionString,
+		in.config.SAContainer,
+		&container.ClientOptions{
+			ClientOptions: azcore.ClientOptions{
+				Cloud: cloud.AzurePublic,
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create blob container pipelineClient: %w", err)
+	}
+
+	// The modern event hub SDK does not create the container
+	// automatically like the old SDK.
+	//
+	// The new `BlobStore` explicitly says:
+	//   "the container must exist before the checkpoint store can be used."
+	//
+	// We need to ensure it exists before we can use it.
+	err = in.ensureContainerExists(ctx, containerClient)
+	if err != nil {
+		return fmt.Errorf("failed to ensure blob container exists: %w", err)
+	}
+
+	// The checkpoint store is used to store the checkpoint information
+	// in the blob container.
+	checkpointStore, err := checkpoints.NewBlobStore(containerClient, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create checkpoint store: %w", err)
+	}
+	in.checkpointStore = checkpointStore
+
+	// Create the event hub consumerClient to receive events.
+	consumerClient, err := azeventhubs.NewConsumerClientFromConnectionString(
+		in.config.ConnectionString,
+		in.config.EventHubName,
+		in.config.ConsumerGroup,
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create consumer pipelineClient: %w", err)
+	}
+	in.consumerClient = consumerClient
+
+	// Manage the migration of the checkpoint information
+	// from the old Event Hub SDK to the new Event Hub SDK.
+	in.migrationAssistant = newMigrationAssistant(
+		in.log,
+		consumerClient,
+		containerClient,
+		checkpointStore,
+	)
+
+	return nil
+}
+
+// run starts the main loop for processing events.
+func (in *eventHubInputV2) run(ctx context.Context) {
+	if in.config.MigrateCheckpoint {
+		in.log.Infow("checkpoint migration is enabled")
+		// Check if we need to migrate the checkpoint store.
+		err := in.migrationAssistant.checkAndMigrate(
+			ctx,
+			in.config.ConnectionString,
+			in.config.EventHubName,
+			in.config.ConsumerGroup,
+		)
+		if err != nil {
+			in.log.Errorw("error migrating checkpoint store", "error", err)
+			// FIXME: should we return here?
```

Comment:
We want to return here?

Reply:
I'm not 100% sure what to do if the migration fails. 

We have two options:

1. Consider it a blocking error and halt the input startup.
2. Consider it a non-blocking error and continue the input startup.

Both scenarios have downsides. 

With no migration:

Scenario 1:
- Users have to check the logs and address migration errors or disable the migration.

Scenario 2:
- The input will start processing log events from the beginning of the event hub retention; some messages may be ingested again.

Reply:
Hmm, got it. Let's then keep it as it is for now. Will revisit this later.